flang/lib/Lower/OpenMP.cpp

```diff
@@ -1710,30 +1710,76 @@ bool ClauseProcessor::processLink(
 createMapInfoOp(fir::FirOpBuilder &builder, mlir::Location loc,
-                mlir::Value baseAddr, std::stringstream &name,
-                mlir::SmallVector<mlir::Value> bounds, uint64_t mapType,
-                mlir::omp::VariableCaptureKind mapCaptureType,
-                mlir::Type retTy) {
-  mlir::Value varPtr, varPtrPtr;
-  mlir::TypeAttr varType;
-
+                mlir::Value baseAddr, mlir::Value varPtrPtr, std::string name,
+                mlir::SmallVector<mlir::Value> bounds,
+                mlir::SmallVector<mlir::Value> members, uint64_t mapType,
+                mlir::omp::VariableCaptureKind mapCaptureType, mlir::Type retTy,
+                bool isVal = false) {
   if (auto boxTy = baseAddr.getType().dyn_cast<fir::BaseBoxType>()) {
     baseAddr = builder.create<fir::BoxAddrOp>(loc, baseAddr);
     retTy = baseAddr.getType();
   }
 
-  varPtr = baseAddr;
-  varType = mlir::TypeAttr::get(
+  mlir::TypeAttr varType = mlir::TypeAttr::get(
       llvm::cast<mlir::omp::PointerLikeType>(retTy).getElementType());
 
   mlir::omp::MapInfoOp op = builder.create<mlir::omp::MapInfoOp>(
-      loc, retTy, varPtr, varType, varPtrPtr, bounds,
+      loc, retTy, baseAddr, varType, varPtrPtr, members, bounds,
       builder.getIntegerAttr(builder.getIntegerType(64, false), mapType),
       builder.getAttr<mlir::omp::VariableCaptureKindAttr>(mapCaptureType),
-      builder.getStringAttr(name.str()));
+      builder.getStringAttr(name));
+
   return op;
 }
 
+static mlir::omp::MapInfoOp processDescriptorTypeMappings(
+    Fortran::semantics::SemanticsContext &semanticsContext,
+    Fortran::lower::StatementContext &stmtCtx,
+    Fortran::lower::AbstractConverter &converter, mlir::Location loc,
+    mlir::Value descriptorAddr, mlir::Value descDataBaseAddr,
+    mlir::ValueRange bounds, std::string asFortran,
+    llvm::omp::OpenMPOffloadMappingFlags mapCaptureType) {
+  llvm::SmallVector<mlir::Value> descriptorBaseAddrMembers;
+  fir::FirOpBuilder &firOpBuilder = converter.getFirOpBuilder();
+
+  mlir::Value descriptor = descriptorAddr;
+
+  // The fir::BoxOffsetOp only works with !fir.ref<!fir.box<...>> types, as
+  // allowing it to access non-reference box operations can cause some
+  // problematic SSA IR. However, in the case of assumed shape's the type
+  // is not a !fir.ref, in these cases to retrieve the appropriate
+  // !fir.ref<!fir.box<...>> to access the data we need to map we must
+  // perform an alloca and then store to it and retrieve the data from the new
+  // alloca.
+  if (fir::isAssumedShape(fir::unwrapRefType(descriptorAddr.getType()))) {
+    mlir::OpBuilder::InsertPoint insPt = firOpBuilder.saveInsertionPoint();
+    firOpBuilder.setInsertionPointToStart(firOpBuilder.getAllocaBlock());
+    descriptor =
+        firOpBuilder.create<fir::AllocaOp>(loc, descriptorAddr.getType());
+    firOpBuilder.restoreInsertionPoint(insPt);
+    firOpBuilder.create<fir::StoreOp>(loc, descriptorAddr, descriptor);
```

Comment:
I am not completely familiar with the lowering pipeline, so can you please clarify where we make sure that the uses of the box inside the region are relinked to the new in-memory box that is actually being mapped?

Reply:
I could be incorrect or misunderstanding your question, but I believe currently with the previous ISA from above changes, all of the target regions symbols are replaced and bound to block arguments into the region (https://github.com/llvm/llvm-project/blob/e50fa9a64f0c37b7e4490bface684e9eff396830/flang/lib/Lower/OpenMP.cpp#L2584), and these block arguments correspond to each map operand (map_info inserted into the map operand list), so these uses are currently relinked in this function effectively when we replace the original descriptor with the newly generated alloca in the final map_info creation below, as the ISA changes effectively changed all of the symbols to block arguments, all that the target region cares about is if we put the right argument with the right type into the corresponding input block argument.

Reply:
There may be cases where we want to rebind the symbol at the current functions scope, but I haven't been able to think of one yet and it'd likely require converting back from a !fir.ref -> to a !fir.box to appease the symbol binding system.

Reply:
I admit I am not completely following what is currently done with the assumed shape arrays, e.g.:
```
subroutine test1(x)
  real :: x(:)
  !$omp target map(to:x)
  !$omp end target
end subroutine test1
```

Gives this FIR:
```
  %5 = omp.map_info var_ptr(%4 : !fir.ref<!fir.array<?xf32>>, !fir.array<?xf32>) map_clauses(to) capture(ByRef) bounds(%3) -> !fir.ref<!fir.array<?xf32>> {name = "x"}
  omp.target map_entries(%5 -> %arg1 : !fir.ref<!fir.array<?xf32>>) {
  ^bb0(%arg1: !fir.box<!fir.array<?xf32>>):
    %6:2 = hlfir.declare %arg1 {uniq_name = "_QFFtest1Ex"} : (!fir.box<!fir.array<?xf32>>) -> (!fir.box<!fir.array<?xf32>>, !fir.box<!fir.array<?xf32>>)
    omp.terminator
  }
```

So there is indeed some misalignment of the types in the `map_entries` and the block arguments.  I suppose after your change the `map_entries` operand will have type `!fir.ref<!fir.box<...`, and this is not worse than it currently is :)

It seems to me the types should align between the two, but this is out of scope of this PR.

Reply:
However, what is currently being done with the assumed shape arrays is creating an allocation, that we can effectively store the passed in box data to and then utilise the box offset op on to retrieve the underlying data pointer, there may be a better way of doing this however, or an extremely obvious way of doing it that I am missing :)